File: 53_Maximum_Subarray.py
# bottom-up DP

class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums)==0:
            return 0
        if len(nums)==1:
            return nums[0]
        maxNum = []
        maxNum.append(nums[0])
        for i in range(len(nums)-1):
            if maxNum[i]>0:
                maxNum.append(maxNum[i]+nums[i+1])
            else:
                maxNum.append(nums[i+1])
        return max(maxNum)

# divide and conquer

class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums)==0:
            return 0
        if len(nums)==1:
            return nums[0]
        low = 0
        high = len(nums)-1
        maxNum, left, right = self.sumMaximum(nums, low, high)
        return maxNum

    def sumMaximum(self, A, low, high):
        if low==high:
            return A[low], low, low
        mid = (low+high)/2
        maxCrossMid, midLow, midHigh = self.maxSubArrayCrossMid(A, mid, low, high)
        maxRight, rightLow, rightHigh = self.sumMaximum(A, mid+1, high)
        maxLeft, leftLow, leftHigh = self.sumMaximum(A, low, mid)
        if maxCrossMid>=maxRight and maxCrossMid>=maxLeft:
            return maxCrossMid, midLow, midHigh
        elif maxRight>=maxLeft and maxRight>=maxCrossMid:
            return maxRight, rightLow, rightHigh
        else:
            return maxLeft, leftLow, leftHigh
               
    def maxSubArrayCrossMid(self, A, mid, low, high):
        sumLeft = float("-inf")  
        left = mid
        curr = 0
        for i in range(mid, low-1, -1):
            curr = curr+A[i]
            if curr>sumLeft:
                sumLeft = curr
                left = i
        sumRight = float("-inf")  
        right = mid+1
        curr = 0
        for i in range(mid+1, high+1, 1):
            curr = curr+A[i]
            if curr>sumRight:
                sumRight = curr
                right = i 
        mSum = sumLeft+sumRight
        return mSum, left, right

# One pass. O(1) space

class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums)==0:
            return 0
        if len(nums)==1:
            return nums[0]
        maxsum = float("-inf")
        currsum = 0
        for i in range(len(nums)):
            currsum = currsum+nums[i]
            maxsum = max(maxsum, currsum)
            if currsum<=0:
                currsum = 0
                
        return maxsum

File: java/53_Maximum_Subarray.java
// Greedy/DP

class Solution {
    public int maxSubArray(int[] nums) {
        int sum = nums[0];
        int max = sum;
        for (int i = 1; i < nums.length; i++) {
            sum = Math.max(nums[i], nums[i]+sum);
            max = Math.max(max, sum);
        }
        return max;
    }
}

// Divide and Conquer

class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 0) return 0;
        if (nums.length == 1) return nums[0];
        return maximumSubArray(nums, 0, nums.length - 1);
    }
    public int maximumSubArray(int[] nums, int low, int high) {
        if (low == high) return nums[low];
        int mid = (low + high) / 2;
        int maxmid = maximumSubArrayAtMid(nums, mid, low, high);
        int maxLeft = maximumSubArray(nums, low, mid);
        int maxRight = maximumSubArray(nums, mid + 1, high);
        return Math.max(maxmid, Math.max(maxLeft, maxRight));
    }
    public int maximumSubArrayAtMid(int[] nums, int mid, int low, int high) {
        int maxL = Integer.MIN_VALUE, curr = 0, maxR = Integer.MIN_VALUE;
        for (int i = mid; i >= low; i--) {
            curr += nums[i];
            maxL = Math.max(curr, maxL);
        }
        curr = 0;
        for (int i = mid + 1; i <= high; i++) {
            curr += nums[i];
            maxR = Math.max(curr, maxR);
        }   
        return maxL + maxR;
    }
}
